### PR TITLE
Handle partial heading copies and id persistence

### PR DESCRIPTION
**Pull request by sbaughman**
It didn't take much to change things to handle partial cut / copy of headings, since CKEditor does the bulk of the work of fixing unpaired HTML tags.  Basically any time it has a partial tag, it will automatically provide the opposing tag, along with attributes.  Because of this, every partially copied heading will already have a copy of the id, so I just made a slight modification to existing logic in `resolveDuplicateIds` to handle it.

Also, in working on the issue of id persistence with changing an element to a non-heading and back, I realized that our main use-case is already covered.  As long as the user doesn't do anything to trigger the Advanced Content Filter (I was originally inadvertently triggering it by viewing source mode), the id will persist with the element, so conversion back to a heading will keep the same id.  I added a test to verify this.

### CODE REVIEW

**Comment by caffodian** [comment]
> As long as the user doesn't do anything to trigger the Advanced Content Filter (I was originally inadvertently triggering it by viewing source mode), the id will persist with the element, so conversion back to a heading will keep the same id. I added a test to verify this.

We can potentially avoid a lot of confusion by just adding `p id` to the filter, then.  Same place we allow headings to have the id. Can you think of any reason not to?

This idea doesn't work for when people have `BR` paragraphs turned on, but we don't use that setting, so it's ok for now.

File: plugins/autoid/plugin.js [hunk 1]
```diff
@@ -139,9 +139,9 @@
           return newHeading;
         }
 
-        // if the text is identical (full copy), the original should retain its
+        // if the original heading text is not blank, the original should retain its
```
Comment (caffodian) [inline plugins/autoid/plugin.js:142]: 👍

**Review by caffodian — APPROVED** [review]
Looks good to me, maybe add p id so that the filter accidentally executing is not a factor?

File: plugins/autoid/plugin.js (hunk 1 above)
Comment (sbaughman) [reply plugins/autoid/plugin.js:142]: I do have a few concerns about enabling paragraph ids:
1. It's not as easy to weed out potential duplication as I originally anticipated.  Doable, but compared to the general simplicity of only allowing ids on headings, noticeably more complex.
2. Ids on paragraphs potentially saved in the documents.  This could cause unexpected behavior if there is ever a future scenario where we find items by id and aren't expecting paragraphs to have ids.  We could filter them out on save by adding a rule to `htmlFilter`, but again, added rules and complexity. 
3.  I think the largest consideration is how this plugin may be used by others.  If the plugin is destined to be open source (I think this one is?  correct me if I'm wrong...), as a potential user I would not expect it to enable id's on anything other than headings.  If we want to make the id persistence more resilient, I would prefer to do it through a more robust mechanism, such as storing current heading information in a variable and doing necessary comparisons  to determine whether a particular id should persist.

Comment (caffodian) [reply plugins/autoid/plugin.js:142]: Okay.  These seem like good reasons not to for now, at least until we better understand usage.  (for example, if people frequently switch back and forth)

Adding it to the filter seems relatively easy in the future, so I am good with leaving this for now.

**Comment by winhamwr** [comment]
TLDR:

> storing current heading information in a variable

What Seth said.

re: paragraph ID's- I think the goal is probably to only care about retaining these during a particular editing session, right?

Instead of storing the old ID in the DOM (and fighting against content filtering), we could store it in javascript via [Expando properties](http://stackoverflow.com/a/1402782/386925). [jquery.data](https://api.jquery.com/jquery.data/) is a tool for doing this while avoiding memory leaks. We could:
- Set a javascript-side expando property when a heading with an ID was converted into something else. We would call e.g. `$(newNonHeadingNode).data('ckeditor-autoid-heading-id', theHeadingId)`.
- Check for the existence of that property when converting a non-heading to a heading, and use the ID if it exists.
- Also store a list of all javascript-side IDs on the editor instance so that we can consider them as part of our uniqueness and copy/paste check. We absolutely must avoid traversing the entire actual DOM to gather these properties for performance reasons, which means we'll need to do a bit of work to keep our global list in sync with the expando attributes on individual elements. Every time we do something that depends on knowing all of the IDs currently in use, we'll need to add the list of javascript-side IDs for consideration.

`jquery.data` exists to prevent memory leaks when setting expando properties, but there's probably a non-jquery solutions for the same problem if we prefer.

**Comment by sbaughman** [comment]
Ah yes... good point on the expando properties.  Also, that reminds me that I totally forgot about  `CKEDITOR.dom.element.setCustomData`, which I believe would accomplish the same thing in the context of an editor session.

**Comment by caffodian** [comment]
Hmm.  This does seem like a possible solution.

The thing to verify is whether or not removing a format style preserves the customData
